File: surf-cloud-standalone/src/main/java/dev/slne/surf/cloud/standalone/Bootstrap.java
package dev.slne.surf.cloud.standalone;

import dev.slne.surf.cloud.api.exceptions.FatalSurfError;
import dev.slne.surf.surfapi.standalone.SurfApiStandaloneBootstrap;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.NestedRuntimeException;

public class Bootstrap {

  @SuppressWarnings("CallToPrintStackTrace")
  public static void main(String[] args) {
    try {
      System.err.println("Classloader: " + Bootstrap.class.getClassLoader());

      SurfApiStandaloneBootstrap.bootstrap();
      final SurfCloudStandaloneInstance instance = SurfCloudStandaloneInstance.get();
      instance.onLoad();
      instance.onEnable();

      Runtime.getRuntime().addShutdownHook(new Thread(() -> {
        instance.onDisable();
        SurfApiStandaloneBootstrap.shutdown();
      }));
    } catch (NestedRuntimeException e) {
      System.err.println("Root cause " + e.getRootCause());
      if (e.getRootCause() instanceof FatalSurfError fatalError) {
        handleFatalError(fatalError);
      } else {
        throw e;
      }
    } catch (FatalSurfError error) {
      handleFatalError(error);
    }
  }

  private static void handleFatalError(FatalSurfError error) {
    System.err.println(error.buildMessage());
    final Throwable cause = error.getCause();

    if (cause != null) {
      cause.printStackTrace();
    }

    final ConfigurableApplicationContext context = SurfCloudStandaloneInstance.get().getDataContext();

    if (context != null && context.isActive()) {
      SpringApplication.exit(context, error);
    } else {
      System.exit(error.exitCode());
    }
  }
}

File: surf-cloud-standalone/src/main/kotlin/dev/slne/surf/cloud/standalone/Bootstrap.kt
package dev.slne.surf.cloud.standalone

import dev.slne.surf.cloud.core.common.CloudCoreInstance
import dev.slne.surf.cloud.core.common.CloudCoreInstance.BootstrapData
import dev.slne.surf.cloud.core.common.coreCloudInstance
import dev.slne.surf.cloud.core.common.handleEventuallyFatalError
import dev.slne.surf.surfapi.core.api.util.logger
import dev.slne.surf.surfapi.standalone.SurfApiStandaloneBootstrap
import kotlinx.coroutines.runBlocking
import org.springframework.boot.SpringApplication
import kotlin.io.path.Path
import kotlin.system.exitProcess

object Bootstrap {
    val log = logger()

    @JvmStatic
    fun main(args: Array<String>): Unit = runBlocking {
        try {
            log.atInfo()
                .log("Classloader: " + Bootstrap::class.java.classLoader)

            SurfApiStandaloneBootstrap.bootstrap()
            SurfApiStandaloneBootstrap.enable()

            coreCloudInstance.bootstrap(
                BootstrapData(
                    dataFolder = Path("")
                )
            )
            coreCloudInstance.onLoad()
            coreCloudInstance.onEnable()
            coreCloudInstance.afterStart()
            log.atInfo()
                .log("Standalone instance is ready!")
        } catch (e: Throwable) {
            e.handleEventuallyFatalError {
                shutdown(it)
            }
        }
    }

    suspend fun shutdown(exitCode: Int): Nothing {
        coreCloudInstance.onDisable()

        val context = CloudCoreInstance.internalContext
        val exitCode =
            if (context != null) SpringApplication.exit(context, { exitCode }) else exitCode
        exitProcess(exitCode)
    }
}